kaiax/gasless/impl/tx_pool.go

```diff
@@ -157,7 +177,56 @@ func (g *GaslessModule) checkBalanceForSwap(swapArgs *SwapArgs, swapNonce uint64
 	return nil
 }
 
-func (g *GaslessModule) IsReady(txs map[uint64]*types.Transaction, i uint64, ready types.Transactions) bool {
+// Check promotion condition and enforce pending pool flow control.
+func (g *GaslessModule) IsReady(txs map[uint64]*types.Transaction, next uint64, ready types.Transactions) bool {
+	g.knownTxsMu.Lock()
+	defer g.knownTxsMu.Unlock()
+
+	tx, ok := txs[next]
+	if !ok {
+		return false
+	}
+
+	if !g.isReady(txs, next, ready) {
+		return false
+	}
+
+	if g.IsBundleTx(tx) {
+		// If prev tx is bundle tx, there's no need to check the knownTxs limit because it has been checked in the previous `IsReady()` execution.
+		isPrevTxBundleTx := len(ready) != 0 && g.IsBundleTx(ready[len(ready)-1])
+		if isPrevTxBundleTx {
+			g.knownTxs.add(tx, TxStatusPending)
+			return true
+		}
+
+		maxBundleTxsInPending := g.GetMaxBundleTxsInPending()
+		if maxBundleTxsInPending != math.MaxUint64 {
+			numExecutable := uint(g.knownTxs.numExecutable())
+
+			numSeqTxs := uint(1)
+			for i := next + 1; i < next+uint64(len(txs)); i++ {
+				if tx, ok := txs[i]; ok && g.IsBundleTx(tx) {
+					numSeqTxs++
+				} else {
+					break
+				}
+			}
+
+			// false if there is possibility of exceeding max bundle tx num
+			if numExecutable+numSeqTxs > maxBundleTxsInPending {
+				logger.Trace("Not promoting a tx because of exceeding max bundle tx num", "tx", tx.Hash().String(), "numExecutable", numExecutable, "maxBundleTxsInPending", maxBundleTxsInPending)
+				return false
+			}
+
+			g.knownTxs.add(tx, TxStatusPending)
+		}
```

Comment:
```suggestion
		}
		g.knownTxs.add(tx, TxStatusPending)
```
If `maxBundleTxsInPending` is MaxUint64, the limitation of pending num is diabled.

Reply:
Thanks. 9eb0787